server/src/app.ts

```diff
@@ -82,17 +78,37 @@ export const main = async (app: Express) => {
           },
         },
       },
+      update: {},
     });
   }
 
+  async function findOrCreateUser(email: string) {
+    try {
+      return await upsertUser(email);
+    } catch (e) {
+      if (
+        e instanceof Prisma.PrismaClientKnownRequestError &&
+        e.code === UNIQUE_CONSTRAINT_FAILED
+      ) {
+        console.error(
+          'Received concurrent login requests - the client is likely making too many requests.',
+        );
```

Comment:
```suggestion
        );
        // Since the error was a unique constraint violation, we know that the
        // user exists, so we can just return their record.
        return await prisma.users.findUniqueOrThrow({ where: { email } });  
```
This _should_ be more efficient because it's definitely a single query.  Also, it's more declarative since we know that we would never be inserting a new record.